### ISSUE TITLE
possible race condition?

### ISSUE THREAD

**Issue by frankurcrazy** [mbrossard/threadpool#10]
Say we create 20 threads with threadpool_create() and call threadpool_add() to add a task into the task queue right away.
Is there possibility that a creating thread (in threadpool_thread()) called pthread_mutex_lock() at line 273 right after threadpool_add() called pthread_cond_signal() at line 186 so that the creating thread acquires the lock before the thread to be woken and in turn, making the pool->count a negative number?

**Comment by mbrossard** [comment]
Hi,

All code that modifies (after first assignment) pool->count does so under lock (of pool->lock).
This might be unclear because `pthread_cond_wait()` releases and reacquires the lock. The call to `pthread_cond_wait()` is wrapped in a loop that will check that count is not zero (line 277), so I see no scenario in which pool->count would become negative.

If you still think there's a problem, write me an example with a sequence of lines in the form "Tx: Ly" where x is the thread number (0/1, 1/2 or a/b whichever you like best), and y is the code line number. You can add a comment with the state of count and lock.

Sincerely,
--
Mathias

**Comment by frankurcrazy** [comment]
Hi,
Thanks for the reply.
Sorry I didn't make it clear. Let me try again.

**main:**
```
pool = threadpool_create(2, 10, 0);
threadpool_add(pool, task1, arg1);
```

Consider there are three threads: **T0**, **T1** and **T2**, T0 is the main thread; T1 is already at pthread_cond_wait() and T2 is just created.

T2:269 -> The thread is just created, haven't enter the for loop
T1:273 -> T1 now acquires the lock
T1:278 -> At this moment, the lock is released
T0:160 -> The lock is acquired by T0 to add task into queue 
T0:186 -> T0 wakes up T1 (not T2, because T2 haven't called ```pthread_cond_wait()``` at the moment) (pool->count = 1)
T2:273 -> T2 tries to acquire lock
T1:278 -> Now T1 wakes up and tries to acquire lock
T0:192 -> T0 unlocks
T2:291 -> **T2 acquires the lock before T1**, and decrements pool->count (pool->count = 0)
T2:294 -> T2 unlocks
T1:278 -> T1 acquires the lock now
T1:291 -> T1 decrements pool->count (pool->count = -1)
T1:294 -> T1 unlocks

**Comment by mbrossard** [comment]
> T1:278 -> T1 acquires the lock now
> T1:291 -> T1 decrements pool->count (pool->count = -1)

That sequence should not happen.

T1:278 -> T1 acquires the lock now [call is in while body; next is 277]
T1:277 -> while condition pool->count == 0 is true [next is while body, 278]
T1:278 -> `pthread_cond_wait()` waits for signal

I had to think about it for a moment. Your message also made me also wonder if it was possible to have a deadlock with a precisely timed `pthread_cond_signal()`: imagine `threadpool_add( )` called when the thread from the thread-pool is between 277 and 278. The only corner case I could spot  was if the thread-pool only has one thread and the main thread waits for the result without making other calls to `threadpool_add( )`.

This only means I should add more comments if only for me.

Sincerely,
Mathias

**Comment by frankurcrazy** [comment]
Yes, you're absolutely right : )
Sorry that I missed the while loop condition even though I've checked many times.
